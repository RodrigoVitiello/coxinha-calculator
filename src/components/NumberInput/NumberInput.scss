.number-input-container {
  display: flex;
  flex-direction: row;
  align-items: center;
  background-color: white;
  border-radius: 5px;
  padding: 10px;
  margin: 5px 0;
  width: 100%;

  label {
    color: gray;
    font-size: 12px;
    flex-grow: 2;
    width: 200px;
  }

  input[type=text] {
    padding: 10px;
    border: 0;
    flex-grow: 99;
    &:focus {
      outline-width: 0;
    }
  }
}
