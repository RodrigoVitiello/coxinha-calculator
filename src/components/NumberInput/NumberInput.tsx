import React, { Component } from 'react';
import './NumberInput.scss'

interface INumberInputProps {
  labelText: string,
  placeholder?: string,
  defaultValue: number,
  handleOnChange: (value: number) => void
}

interface INumberInputState {
  currentValue: number
}

class NumberInput extends Component<INumberInputProps, INumberInputState> {
  constructor(props: INumberInputProps) {
    super(props)

    this.state = {
      currentValue: 0
    }
  }

  componentDidMount() {
    this.setState({currentValue: this.props.defaultValue})
  }

  async handleInputValue(value: string) {
    if (value.match(/^\d+$/g)) {
      const newValue = Number(value)
      await this.setState({currentValue: newValue})
    } else if (value.length === 0) {
      await this.setState({currentValue: 0})
    }

    this.props.handleOnChange(this.state.currentValue)
  }

  render() {
    const { labelText, placeholder } = this.props
    const labelId: string = `number-label-${Math.random().toString(36).replace(/[^a-z]+/g, '')}`

    return (
      <div className='number-input-container'>
        <label htmlFor={labelId}>{labelText}</label>
        <input
          id={labelId}
          aria-label='number-input'
          type='text'
          value={this.state.currentValue}
          onChange={ (e) => this.handleInputValue(e.target.value)}
          placeholder={placeholder}
        />
      </div>
    )
  }
}

export default NumberInput;
