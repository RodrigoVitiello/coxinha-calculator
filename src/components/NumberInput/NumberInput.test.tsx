import React from 'react'
import { render, fireEvent } from '@testing-library/react'
import NumberInput from './NumberInput'

describe('NumberInput', () => {
  const handleOnChangeMock = jest.fn()
  const defaultValue: number = Math.floor(Math.random() * 100)
  const component = (
    <NumberInput
      defaultValue={defaultValue}
      labelText='my input'
      handleOnChange={handleOnChangeMock}
    />
  )
  const inputQuery: string = '.number-input-container input'

  xit('renders number input component', () => {
    const { container } = render(component);

    expect(container).toMatchSnapshot();
  });

  it('renders the default values correctly', () => {
    const { container } = render(component)
    const label = container.querySelector('.number-input-container label')
    const input = container.querySelector(inputQuery)

    expect(label).toHaveTextContent('my input')
    expect(input).toHaveValue(String(defaultValue))
  })

  describe('changing the field value', () => {
    const newInputValue: string = String(Math.floor(Math.random() * 100))

    it('triggers callback on input change', async () => {
      const { container } = render(component)
      const input = container.querySelector(inputQuery)

      await fireEvent.change(input, { target: { value: newInputValue } })

      expect(handleOnChangeMock.mock.calls.length).toBe(1)
    })

    it('triggers callback with the new value converted to type number', async () => {
      const { container } = render(component)
      const input = container.querySelector(inputQuery)
      const expectedValue: number = Number(newInputValue)
      await fireEvent.change(input, { target: { value: newInputValue } })

      expect(handleOnChangeMock.mock.calls[0][0]).toBe(expectedValue)
    })

    it('updates the input value', () => {
      const { container } = render(component)
      const input = container.querySelector(inputQuery)

      fireEvent.change(input, { target: { value: newInputValue } })

      expect(input).toHaveValue(newInputValue)
    })

    it('does not accept invalid characters and keep the previous value', () => {
      const { container } = render(component)
      const input = container.querySelector(inputQuery)

      expect(input).toHaveValue(String(defaultValue))
      fireEvent.change(input, { target: { value: 'bla' } })
      expect(input).toHaveValue(String(defaultValue))
    })

    it('keep value 0 when deleting all characters from input', () => {
      const { container } = render(component)
      const input = container.querySelector(inputQuery)

      expect(input).toHaveValue(String(defaultValue))
      fireEvent.change(input, { target: { value: '' } })
      expect(input).toHaveValue('0')
    })
  })
})
