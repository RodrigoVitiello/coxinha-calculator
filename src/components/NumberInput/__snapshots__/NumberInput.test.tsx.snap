// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput renders number input component 1`] = `
<div>
  <div
    class="number-input-container"
  >
    <label
      for="number-label-qdkvzofo"
    >
      my input
    </label>
    <input
      aria-label="number-input"
      id="number-label-qdkvzofo"
      type="text"
      value="15"
    />
  </div>
</div>
`;
